govcd/catalog.go

```diff
@@ -55,3 +66,243 @@ func (c *Catalog) FindCatalogItem(catalogitem string) (CatalogItem, error) {
 
 	return CatalogItem{}, fmt.Errorf("can't find catalog item: %s", catalogitem)
 }
+
+//add callback or channel
+func (c *Catalog) UploadOvf(ovaFileName, itemName, description string, chunkSize int) error {
+
+	catalogItemUploadURL, err := findCatalogItemUploadLink(c)
+	if err != nil {
+		return err
+	}
+
+	vappTemplateUrl, err := createItemForUpload(c.c, catalogItemUploadURL, itemName, description)
+	if err != nil {
+		return err
+	}
+
+	vAppTemplate, err := queryVappTemplate(c.c, vappTemplateUrl)
+	if err != nil {
+		return err
+	}
+
+	ovfUploadHref, err := getOvfUploadlink(vAppTemplate)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("AAA:", ovfUploadHref)
+
+	filesAbsPaths, err := Unpack(ovaFileName)
+	if err != nil {
+		return err
+	}
+
+	for _, path := range filesAbsPaths {
+		if filepath.Ext(path) == ".ovf" {
+			err := uploadOvfDescription(c.c, path, ovfUploadHref)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	for {
+		log.Printf("[TRACE] Sleep... for 5 seconds.\n")
+		time.Sleep(time.Second * 5)
+		vAppTemplate, err = queryVappTemplate(c.c, vappTemplateUrl)
+		if err != nil {
+			return err
+		}
+		if len(vAppTemplate.Files.File) > 1 {
+			log.Printf("[TRACE] upload link prepared.\n")
+			break
+		}
+	}
+
+	err = uploadFiles(c.c, filesAbsPaths, *vAppTemplate)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[TRACE] End ^^..^^ \n")
+	return nil
+}
+
+func getOvfUploadlink(vappTemplate *types.VAppTemplate) (*url.URL, error) {
+	log.Printf("[TRACE] Parsing ofv upload link: %#v\n", vappTemplate)
+
+	ovfUploadHref, err := url.ParseRequestURI(vappTemplate.Files.File[0].Link[0].HREF)
+	if err != nil {
+		return nil, err
+	}
+
+	return ovfUploadHref, nil
+}
+
+func queryVappTemplate(client *Client, vappTemplateUrl *url.URL) (*types.VAppTemplate, error) {
+	log.Printf("[TRACE] Qeurying vapp template: %s\n", vappTemplateUrl)
+	request := client.NewRequest(map[string]string{}, "GET", *vappTemplateUrl, nil)
+	response, err := client.Http.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	vAppTemplateParsed := &types.VAppTemplate{}
+
+	err = xml.Unmarshal(body, &vAppTemplateParsed)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Printf("[TRACE] Response: %#v\n", response)
```

Comment:
Please note that PR #76 will log HTTP requests and responses automatically, if you use `NewRequest` and `decodeBody` from api.go.

Reply:
Nice. Now it's racing ;-D